File: projects/ruby/lib/operations.rb
# frozen_string_literal: true

require 'securerandom'
require 'time'

##
# Operations module provides handlers for all DAO operations.
# Each handler processes params and returns a result payload.
#
module Operations
  module_function

  ##
  # Process a prompt with simulated AI processing.
  #
  # @param params [Hash] Parameters with :text key
  # @return [Hash] Processed prompt result
  # @raise [ArgumentError] If text is missing or invalid
  #
  def process_prompt(params)
    text = params['text'] || params[:text]
    raise ArgumentError, 'Missing required parameter: text' if text.nil? || text.empty?
    raise ArgumentError, 'Text must be between 1 and 10000 characters' if text.length > 10_000

    # Simulated AI processing: uppercase + suffix
    {
      text: text,
      processed: "#{text.upcase} [AI-processed]",
      timestamp: Time.now.utc.iso8601
    }
  end

  ##
  # Create a reusable template with variable placeholders.
  #
  # @param params [Hash] Parameters with :name, :pattern, :variables keys
  # @return [Hash] Created template with UUID
  # @raise [ArgumentError] If required params are missing or invalid
  #
  def create_template(params)
    name = params['name'] || params[:name]
    pattern = params['pattern'] || params[:pattern]
    variables = params['variables'] || params[:variables] || []

    raise ArgumentError, 'Missing required parameter: name' if name.nil? || name.empty?
    raise ArgumentError, 'Missing required parameter: pattern' if pattern.nil? || pattern.empty?
    raise ArgumentError, 'Invalid template name: must be alphanumeric with hyphens' unless name.match?(/^[a-z0-9-]+$/i)

    {
      id: SecureRandom.uuid,
      name: name,
      pattern: pattern,
      variables: variables
    }
  end

  ##
  # Render a template with variable substitution.
  #
  # @param params [Hash] Parameters with :template_id, :values keys
  # @return [Hash] Rendered template result
  # @raise [ArgumentError] If required params are missing
  #
  def render_template(params)
    template_id = params['template_id'] || params[:template_id]
    values = params['values'] || params[:values] || {}

    raise ArgumentError, 'Missing required parameter: template_id' if template_id.nil?
    raise ArgumentError, 'Template not found' unless valid_uuid?(template_id)

    # For demo: simple pattern substitution
    # In real implementation, would lookup template from store
    rendered = "Rendered template #{template_id} with variables"

    {
      template_id: template_id,
      rendered: rendered,
      variables_used: values
    }
  end

  ##
  # Track usage of an operation for analytics.
  #
  # @param params [Hash] Parameters with :operation, :duration_ms keys
  # @return [Hash] Tracking confirmation
  # @raise [ArgumentError] If required params are missing or invalid
  #
  def track_usage(params)
    operation = params['operation'] || params[:operation]
    duration_ms = params['duration_ms'] || params[:duration_ms]

    raise ArgumentError, 'Missing required parameter: operation' if operation.nil? || operation.empty?
    raise ArgumentError, 'Missing required parameter: duration_ms' if duration_ms.nil?
    raise ArgumentError, 'duration_ms must be non-negative' if duration_ms.to_i.negative?

    {
      tracked: true,
      operation: operation,
      timestamp: Time.now.utc.iso8601
    }
  end

  ##
  # Validate UUID format.
  #
  # @param uuid [String] UUID string to validate
  # @return [Boolean] True if valid UUID v4 format
  #
  def valid_uuid?(uuid)
    uuid.match?(/^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i)
  end
end

File: projects/javascript/lib/operations.js
/**
 * Operations module - Handlers for all DAO operations
 *
 * Each function processes params and returns a result payload.
 */

import { v4 as uuidv4 } from 'uuid';

/**
 * Process a prompt with simulated AI processing
 * @param {object} params - Parameters with text field
 * @returns {object} Processed prompt result
 */
export function processPrompt(params) {
  const text = params.text;

  if (!text || typeof text !== 'string' || text.length === 0) {
    throw new Error('Missing required parameter: text');
  }

  if (text.length > 10000) {
    throw new Error('Text must be between 1 and 10000 characters');
  }

  // Simulated AI processing: uppercase + suffix
  return {
    text: text,
    processed: `${text.toUpperCase()} [AI-processed]`,
    timestamp: new Date().toISOString()
  };
}

/**
 * Create a reusable template with variable placeholders
 * @param {object} params - Parameters with name, pattern, variables
 * @returns {object} Created template with UUID
 */
export function createTemplate(params) {
  const { name, pattern, variables = [] } = params;

  if (!name || typeof name !== 'string' || name.length === 0) {
    throw new Error('Missing required parameter: name');
  }

  if (!pattern || typeof pattern !== 'string' || pattern.length === 0) {
    throw new Error('Missing required parameter: pattern');
  }

  // Validate name format (alphanumeric + hyphens)
  if (!/^[a-zA-Z0-9-]+$/.test(name)) {
    throw new Error('Invalid template name: must be alphanumeric with hyphens');
  }

  return {
    id: uuidv4(),
    name: name,
    pattern: pattern,
    variables: variables
  };
}

/**
 * Render a template with variable substitution
 * @param {object} params - Parameters with template_id, values
 * @returns {object} Rendered template result
 */
export function renderTemplate(params) {
  const { template_id, values = {} } = params;

  if (!template_id || typeof template_id !== 'string') {
    throw new Error('Missing required parameter: template_id');
  }

  // For demo: simple rendered output
  const rendered = `Rendered template ${template_id} with variables`;

  return {
    template_id: template_id,
    rendered: rendered,
    variables_used: values
  };
}

/**
 * Track usage of an operation for analytics
 * @param {object} params - Parameters with operation, duration_ms
 * @returns {object} Tracking confirmation
 */
export function trackUsage(params) {
  const { operation, duration_ms } = params;

  if (!operation || typeof operation !== 'string' || operation.length === 0) {
    throw new Error('Missing required parameter: operation');
  }

  if (duration_ms === undefined || duration_ms === null) {
    throw new Error('Missing required parameter: duration_ms');
  }

  if (typeof duration_ms !== 'number' || duration_ms < 0) {
    throw new Error('duration_ms must be non-negative');
  }

  return {
    tracked: true,
    operation: operation,
    timestamp: new Date().toISOString()
  };
}
